### PR TITLE
Add addcom skeleton

### PR DESCRIPTION
**Pull request by zlimez**
- [X] Addcom command parser
- [X] Commission Builder (Optional description)
- [X] UI updates to show added commission
- [X] Verbose feedback for added commission

Concerns
- Preferably not use replacement mechanism for customer and opt for update instead? (Amar's suggestion of a status counter in the observable list?)
- HashSet does not preserve ordering of adding commissions
- The commission list to customer data structure not utilised

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/AddCommissionCommand.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            + "Parameters: "
+            + PREFIX_TITLE + "TITLE "
+            + PREFIX_FEE + "FEE "
+            + PREFIX_DEADLINE + "DEADLINE "
+            + "[" + PREFIX_DESCRIPTION + "DESCRIPTION] "
+            + "[" + PREFIX_TAG + "TAG]...\n"
+            + "Example: " + COMMAND_WORD + " "
+            + PREFIX_TITLE + "Tokyo Ghoul Kaneki "
+            + PREFIX_FEE + "50 "
+            + PREFIX_DEADLINE + "johnd@example.com "
```
Comment (ngxingyu) [inline src/main/java/seedu/address/logic/commands/AddCommissionCommand.java:30]: Can you update the command usage message so we know what parameters the `addcom` command accepts? Thanks.

Comment (seelengxd) [reply src/main/java/seedu/address/logic/commands/AddCommissionCommand.java:30]: yes please update it :P 

`addcom n/test f/50 d/2022-02-02 s/true p/desc t/digital t/neon palette`

but tag regex checks currently don't allow spaces. we would need to enable it for the second tag

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 2]
```diff
@@ -95,6 +103,88 @@ public static Email parseEmail(String email) throws ParseException {
...
+     */
+    public static Description parseDescription(String description) {
+        requireNonNull(description);
+        String trimmedDescription = description.trim();
+        return new Description(trimmedDescription);
+    }
+
+
+
+
```
Comment (seelengxd) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:187]:
```suggestion
```
Should we remove all these new lines?

File: src/main/java/seedu/address/model/ModelManager.java [hunk 3]
```diff
@@ -138,8 +138,9 @@ public void deleteCommission(Commission target) {
 
     @Override
     public void addCommission(Commission commission) {
-        addressBook.addCommission(getSelectedCustomer().getValue(), commission);
-        updateFilteredCustomerList(PREDICATE_SHOW_ALL_CUSTOMERS);
+        Customer updatedCustomer = addressBook.addCommission(getSelectedCustomer().getValue(), commission);
+        selectCustomer(updatedCustomer);
+        // filteredCommissions = new FilteredList<>(getCommissionList());
```
Comment (seelengxd) [inline src/main/java/seedu/address/model/ModelManager.java:143]:
```suggestion
```

File: src/main/java/seedu/address/model/commission/Commission.java [hunk 4]
```diff
@@ -139,4 +140,41 @@ public String toString() {
         return builder.toString();
     }
 
+    /**
+     * Builder class for Commission.
+     */
+    public static class CommissionBuilder {
```
Comment (seelengxd) [inline src/main/java/seedu/address/model/commission/Commission.java:146]: Shall we add the null checks back for CommissionBuilder? (yes i realise i also forgot)

so something like this https://github.com/AY2223S1-CS2103T-W11-3/tp/pull/79

**Review by seelengxd — DISMISSED** [review]
Found 2 more bugs to fix.

File: src/main/java/seedu/address/logic/commands/AddCommissionCommand.java (hunk 1 above)
Comment (zlimez) [reply src/main/java/seedu/address/logic/commands/AddCommissionCommand.java:30]: Kk soz missed it

File: src/main/java/seedu/address/logic/commands/AddCommissionCommand.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+     */
+    public AddCommissionCommand(Commission commission) {
+        requireNonNull(commission);
+        toAdd = commission;
+    }
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        requireNonNull(model);
+
```
[the remark this answers is not in the document]
Comment (zlimez) [reply src/main/java/seedu/address/logic/commands/AddCommissionCommand.java:51]: okie i will fix it now

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 6]
```diff
@@ -51,7 +59,7 @@ public static Name parseName(String name) throws ParseException {
     }
 
     /**
-     * Parses a {@code String phone} into a {@code Phone}.
+     * Parses a {@code String phone} into a {@code Fee}.
```
Comment (ngxingyu) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:62]:
```suggestion
     * Parses a {@code String phone} into a {@code Phone}.
```

File: src/main/java/seedu/address/model/AddressBook.java [hunk 7]
```diff
@@ -121,20 +121,33 @@ public void removeCustomer(Customer key) {
...
         requireAllNonNull(customer, commission);
+        // Hashset would cause the commissions to be out of order
         HashSet<Commission> customerCommissions = new HashSet<>(getCustomerCommissions(customer));
         customerCommissions.add(commission);
+        commission.setCustomer(customer);
         commissions.add(commission);
+
         Customer newCustomer = customer.copyWithCommissions(customerCommissions);
+        // Not efficient but ensures all commissions have reference to the new customer who replaced the old one
+        // More ideal to keep same customer reference and update the object instead @Amar suggestion to add counter
```
Comment (ngxingyu) [inline src/main/java/seedu/address/model/AddressBook.java:143]: I agree with this point: we shouldn't need to constantly update the entire list of commissions just to trigger the redraw of the UI.

**Review by ngxingyu — CHANGES_REQUESTED** [review]
Only 1 major issue - `commission.setCustomer` should not be called within `hasCommission` since it is not guaranteed that the commission belongs to the customer being called. (It works for all existing uses of `hasCommission` but might break in the future)

File: src/main/java/seedu/address/model/AddressBook.java (hunk 7 above)
Comment (seelengxd) [reply src/main/java/seedu/address/model/AddressBook.java:143]: kinda random (also haven't tested or tried using it and probably not necessary yet in this PR) but apparently `ListView` (the thing `CustomerListPanel` has that uses the `observableList` has this interesting method)

## `ListView.refresh()`

```
 * Calling {@code refresh()} forces the ListView control to recreate and
 * repopulate the cells necessary to populate the visual bounds of the control.
 * In other words, this forces the ListView to update what it is showing to
 * the user. This is useful in cases where the underlying data source has
 * changed in a way that is not observed by the ListView itself.
```

File: src/main/java/seedu/address/model/AddressBook.java [hunk 8]
```diff
@@ -121,20 +121,33 @@ public void removeCustomer(Customer key) {
      */
     public boolean hasCommission(Customer customer, Commission commission) {
         requireAllNonNull(customer, commission);
+        // Previously duplicate commission won't be detected because customer field has not been set.
+        commission.setCustomer(customer);
```
[the remark this answers is not in the document]
Comment (zlimez) [reply src/main/java/seedu/address/model/AddressBook.java:125]: Yeah, I agree this looks disgusting but if I don't set the customer here duplicate commission will not be detected because the equals method involves comparing the customer field as well so will cause nullpointerexception. when the duplicate is later detected in during the HashSet add. I did this because for now isSameCommission() only compares the title and not the customer. Else I can move the set customer to the addcommission in ModelManager which I think will prevent this issue where the selected customer is the owner of the commission. Just that it seems to break the abstraction layers?

File: data/addressbook.json [hunk 9]
```diff
@@ -12,6 +12,12 @@
       "deadline" : "2022-04-04",
       "tagged" : [ "test2", "test" ],
       "isCompleted" : true
+    }, {
+      "title" : "Berserk",
+      "description" : "",
+      "fee" : 100.0,
+      "deadline" : "2022-10-10",
+      "tagged" : [ "inking" ],
```
[the remark this answers is not in the document]
Comment (zlimez) [reply data/addressbook.json:20]: LOL I accidentally deleted it when resolving merge conflict haizz

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 10]
```diff
@@ -95,6 +103,89 @@ public static Email parseEmail(String email) throws ParseException {
...
+    /**
+     * Parses a {@code String deadline} into a {@code Deadline}.
+     * Leading and trailing whitespaces will be trimmed.
+     *
+     * @throws ParseException if the given {@code deadline} is invalid.
+     */
+    public static CompletionStatus parseStatus(String status) throws ParseException {
+        requireNonNull(status);
+        String trimmedStatus = status.trim();
+        if (trimmedStatus.equalsIgnoreCase("True") || trimmedStatus.equalsIgnoreCase("T")
```
[the remark this answers is not in the document]
Comment (zlimez) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java:168]: Hmm sure but that would require the same checks to be done twice? One for validity and one for actually parsing which status it is (duplicate code)? Or is there a better way? Moving the Me

File: src/main/java/seedu/address/model/AddressBook.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ngxingyu) [reply src/main/java/seedu/address/model/AddressBook.java:125]: My main concern with this approach is the possibility that `hasCommission` changes the customer of commission wrongly if the commission belongs to a different customer, I think setting the customer within addcommission makes more sense since it is supposed to add the commission to the customer's commission list, and after it is added, the commission's customer should be the customer that it was just added to.

File: src/main/java/seedu/address/logic/parser/ParserUtil.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (ngxingyu) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java:168]: Perhaps we can have a regex for completed and a regex for not completed so the validity check (e.g. [VALIDATION_REGEX](https://github.com/AY2223S1-CS2103T-W11-3/tp/blob/master/src/main/java/seedu/address/model/customer/Email.java#L32)) can just be a combination of both

File: src/main/java/seedu/address/logic/commands/DeleteCommissionCommand.java [hunk 11]
```diff
@@ -48,8 +47,6 @@ public CommandResult execute(Model model) throws CommandException {
         Commission commissionToDelete = lastShownList.get(targetIndex.getZeroBased());
         model.deleteCommission(commissionToDelete);
 
-        model.updateFilteredCustomerList(PREDICATE_SHOW_ALL_CUSTOMERS);
-
```
Comment (zlimez) [inline src/main/java/seedu/address/logic/commands/DeleteCommissionCommand.java:52]: I think this is unnecessary?

File: src/main/java/seedu/address/model/AddressBook.java [hunk 12]
```diff
@@ -162,13 +172,14 @@ public void removeCommission(Customer customer, Commission key) {
...
+        //Set<Commission> filteredCommissions = getCustomerCommissions(customer).stream()
+        //        .filter(commission -> !commission.equals(key)).collect(Collectors.toSet());
         commissions.remove(key);
         // make a copy of customer with the updated commissions.
-        Customer newCustomer = customer.copyWithCommissions(filteredCommissions);
+        // Customer newCustomer = customer.copyWithCommissions(filteredCommissions);
         // replace original customer with updated customer to trigger customer list UI updates
-        setCustomer(customer, newCustomer);
+        // setCustomer(customer, newCustomer);
+        customer.getCommissionsMutable().remove(key);
```
Comment (zlimez) [inline src/main/java/seedu/address/model/AddressBook.java:182]: Hmm, I modified this because originally it will cause a bug which after deleting commission A, then subsequently trying to add commission A a duplicate commission error is shown. I think it is because selectedCustomer still references the old customer even though a new customer replaced the instance. I don't think it is necessary to replace a customer with the deletion command since the observable list can detect an object being removed. More efficient also? 

P.S. I added another getter in customer class to enable mutability, don't know if that's the best.

File: src/main/java/seedu/address/logic/parser/ParserUtil.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (zlimez) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java:168]: Ok sure give me a moment then

File: src/main/java/seedu/address/model/AddressBook.java (hunk 12 above)
Comment (ngxingyu) [reply src/main/java/seedu/address/model/AddressBook.java:182]: There might be issues with exposing the commission list of the customer like this in the long run since this would allow anyone to edit it freely, but it can be fixed in #82 when `removeCommission` is moved into `Customer`

**Review by ngxingyu — APPROVED** [review]
Overall LGTM, the addcom command works as described. Only the `getCommissionsMutable` method might introduce bugs in future, but this will be fixed in #76 so I'm fine with merging this.

**Review by seelengxd — APPROVED** [review]
LGTM! Leaving commented code in `AddressBook.java` seems slightly scuffed, but I assume that can also be fixed in #82. Thanks for fixing the wrongly copy-pasted javadocs! :P
